core/modules/memman/MemFileSet.cc

```diff
@@ -0,0 +1,136 @@
+/******************************************************************************/
+  
+MemFileSet::~MemFileSet() {
+
+    // Unreference every fle in our file set. This action will also cause
+    // memory to be unlocked if no one else is using the file then the file
+    // object will be deleted as well.
+    //
+    for (auto mfP : _lockFiles) {mfP->release();}
+    for (auto mfP : _flexFiles) {mfP->release();}
+}
+
+/******************************************************************************/
+/*                                   a d d                                    */
+/******************************************************************************/
+
+int MemFileSet::add(std::string const& tabname, int chunk,
+                    bool iFile, bool mustLK) {
+
+std::string fPath(_memory.filePath(tabname, chunk, iFile));
+
+    // Obtain a memory file object for this table and chunk
+    //
+    MemFile::MFResult mfResult = MemFile::obtain(fPath, _memory);
+    if (mfResult.mfP == 0) return mfResult.retc;
+
+    // Add to the appropriate file set
+    //
+    if (mustLK) _lockFiles.push_back(mfResult.mfP);
+       else     _flexFiles.push_back(mfResult.mfP);
+    _numFiles++;
+    return 0;
+}
+  
+/******************************************************************************/
+/*                               l o c k A l l                                */
+/******************************************************************************/
+
+int MemFileSet::lockAll() {
+
+MemFile::MLResult mlResult;
+size_t bytesLocked, bytesMax, freeBytes;
+
+    // Calculate the number of bytes available at this point. Note that we force
+    // freeBytes to be atleast 1 to make memlock check before it tries locking
+    // the file to avoid a useless memory map operation if it can't be locked.
+    // By the time we get here someone else may have already locked the file.
+    //
+    bytesMax    = _memory.bytesMax();
+    bytesLocked = _memory.bytesLocked();
+    if (bytesMax <= bytesLocked) freeBytes = 1;
+       else freeBytes = bytesMax - bytesLocked;
+
+    // Try to lock all of the required tables
+    //
+    for (auto mfP : _lockFiles)
+        {mlResult = mfP->memLock(freeBytes);
+         if (mlResult.retc != 0) return mlResult.retc;
+         _lockBytes += mlResult.bLocked;
+         if (freeBytes > mlResult.bLocked) freeBytes -= mlResult.bLocked;
+            else freeBytes = 1;
+        }
+
+    // Try locking as many flexible files as we can. We only lock the file table
+    // if the reference count >= 2 to optimize memory usage. At some point we
+    // will place unlocked flex files on a "want to lock" queue. FUTURE!!!
+    //
+    for (auto mfP : _flexFiles)
+        {mlResult = mfP->memLock(freeBytes, 2);
+         if (mlResult.bLocked == 0) continue;
+         _lockBytes += mlResult.bLocked;
+         if (freeBytes > mlResult.bLocked) freeBytes -= mlResult.bLocked;
+            else freeBytes = 1;
+        }
```

Comment:
Looking at this, I think we want to track how often flex lock fails to lock the file. If it happens a lot, we will want to know.